test/gtest/ucp/test_ucp_proto_select.cc

```diff
@@ -0,0 +1,447 @@
+    {
+        uct_tl_t *tl;
+        ucs_list_for_each(tl, &m_component->tl_list, list) {
+            if (0 == strcmp(tl->name, params->mode.device.tl_name)) {
+                return tl;
+            }
+        }
+
+        return nullptr;
+    }
+
+    static ucs_status_t iface_open_mock(uct_md_h md, uct_worker_h worker,
+                                        const uct_iface_params_t *params,
+                                        const uct_iface_config_t *config,
+                                        uct_iface_h *iface_p)
+    {
+        mock_component *self = find(md->component);
+        uct_tl_t *tl         = self->find_tl(params);
+        ucs_status_t status  = mock_t::actual_call(&tl->iface_open, md, worker,
+                                                   params, config, iface_p);
+        if (status == UCS_OK) {
+            uct_base_iface_t *base = ucs_derived_of(*iface_p, uct_base_iface_t);
+            self->m_iface_names[base] = std::string(params->mode.device.tl_name)
+                                        + "/" + params->mode.device.dev_name;
+
+            mock_t::setup(&base->internal_ops->iface_estimate_perf,
+                          iface_estimate_perf_mock);
+            mock_t::setup(&(*iface_p)->ops.iface_query, iface_query_mock);
+        }
+
+        return status;
+    }
+
+    static ucs_status_t
+    iface_estimate_perf_mock(uct_iface_h iface, uct_perf_attr_t *perf_attr)
+    {
+        uct_base_iface_t *base = ucs_derived_of(iface, uct_base_iface_t);
+        ucs_status_t status    =
+            mock_t::actual_call(&base->internal_ops->iface_estimate_perf, iface,
+                                perf_attr);
+        if (status == UCS_OK) {
+            mock_component *self = find(base->md->component);
+            auto it = self->m_perf_attrs_funcs.find(self->m_iface_names[base]);
+            if (it != self->m_perf_attrs_funcs.end()) {
+                (it->second)(*perf_attr);
+            }
+        }
+
+        return status;
+    }
+
+    static ucs_status_t
+    iface_query_mock(uct_iface_h iface, uct_iface_attr_t *iface_attr)
+    {
+        ucs_status_t status =
+            mock_t::actual_call(&iface->ops.iface_query, iface, iface_attr);
+        if (status == UCS_OK) {
+            uct_base_iface_t *base = ucs_derived_of(iface, uct_base_iface_t);
+            mock_component *self   = find(base->md->component);
+            auto it = self->m_iface_attrs_funcs.find(self->m_iface_names[base]);
+            if (it != self->m_iface_attrs_funcs.end()) {
+                (it->second)(*iface_attr);
+            }
+        }
+
+        return status;
+    }
+
+    uct_component_h                           m_component;
+    uint64_t                                  m_flags_orig;
+    std::map<uct_base_iface_t *, std::string> m_iface_names;
+    std::map<std::string, iface_attr_func_t>  m_iface_attrs_funcs;
+    std::map<std::string, perf_attr_func_t>   m_perf_attrs_funcs;
```

Comment:
Is there a necessity to be able to set different values for different transport devices? WDYTA removing all this maps and replacing them by one instance of `iface_attr_func_t` and `perf_attr_func_t` which would be used to mock all the corresponding calls?

Reply:
It's hard to say whether it's really needed, it depends what kind of tests we wish to create.
This gives us possibility to mock multiple transport/ifaces within the same test, which I think might be useful to:
1. Test proto selection between multiple ifaces (e.g. when we switch from one iface to another)
2. Test lane selection
I thought it's important to have this possibility (especially if it does not come with huge complexity). Only if we are 100% sure that those tests will never be needed - then we can simplify and stick to a single instance.